lib/Sema/CSDiagnostics.cpp

```diff
@@ -2352,3 +2352,133 @@ bool InaccessibleMemberFailure::diagnoseAsError() {
   emitDiagnostic(Member, diag::decl_declared_here, Member->getFullName());
   return true;
 }
+
+bool ImplicitCoercionToAnyFailure::diagnoseAsError() {
+  auto E = getAnchor();
+  auto subExpr = E->getValueProvidingExpr();
+  auto &TC = getTypeChecker();
+  Decl *decl = nullptr;
+
+  if (auto *UDE = dyn_cast<UnresolvedDotExpr>(E)) {
+    auto name = UDE->getName();
+    auto parent = getDC()->getParentForLookup();
+    SmallVector<ValueDecl *, 4> lookupResults;
+    parent->lookupQualified(FromType->lookThroughAllOptionalTypes(), name,
+                            NL_QualifiedDefault, nullptr, lookupResults);
+    if (!lookupResults.empty()) {
+      decl = lookupResults.front();
+    }
+  }
+
+  auto hasImplicitlyUnwrappedResult = [&](Expr *E) -> bool {
+    auto getDeclForExpr = [&](Expr *E) -> ValueDecl * {
+      if (auto *call = dyn_cast<CallExpr>(E))
+        E = call->getDirectCallee();
+
+      if (auto *subscript = dyn_cast<SubscriptExpr>(E)) {
+        if (subscript->hasDecl())
+          return subscript->getDecl().getDecl();
+
+        return nullptr;
+      }
+
+      if (auto *memberRef = dyn_cast<MemberRefExpr>(E))
+        return memberRef->getMember().getDecl();
+      if (auto *declRef = dyn_cast<DeclRefExpr>(E))
+        return declRef->getDecl();
+      if (auto *apply = dyn_cast<ApplyExpr>(E))
+        return apply->getCalledValue();
+
+      return nullptr;
+    };
+
+    // Look through implicit conversions like loads, derived-to-base
+    // conversion, etc.
+    if (auto *ICE = dyn_cast<ImplicitConversionExpr>(E))
+      E = ICE->getSubExpr();
+
+    auto *decl = getDeclForExpr(E);
+
+    return decl &&
+           decl->getAttrs().hasAttribute<ImplicitlyUnwrappedOptionalAttr>();
+  };
+
+  auto isOptionalToAnyCoercion = [&](Type srcType, Type destType,
+                                     size_t &difference) -> bool {
+    SmallVector<Type, 4> destOptionals;
+    auto destValueType = destType->lookThroughAllOptionalTypes(destOptionals);
+
+    if (!destValueType->isAny())
+      return false;
+
+    SmallVector<Type, 4> srcOptionals;
+    srcType->lookThroughAllOptionalTypes(srcOptionals);
+
+    if (srcOptionals.size() > destOptionals.size()) {
+      difference = srcOptionals.size() - destOptionals.size();
+      return true;
+    } else {
+      return false;
+    }
+  };
+
+  auto emitSilenceOptionalAnyWarningWithCoercion = [&](Expr *E, Type destType) {
```

Comment:
Too bad this logic as well as `isOptionalToAnyCoercion`  has to be duplicated between MiscDiagnostics and here... It would be great to find a better place for this.